resources/assets/lib/gallery-contest-vote-progress.tsx

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import { CircularProgress } from 'circular-progress';
+import * as React from 'react';
+
+type Props = Record<string, never>;
+
+interface State {
+  maxVotes: number;
+  voteCount: number;
+}
+
+export default class GalleryContestVoteProgress extends React.PureComponent<Props, State> {
+  constructor(props: Props) {
+    super(props);
+
+    this.state = this.getVoteState();
+  }
+
+  componentDidMount() {
+    $.subscribe('contest:vote:end.count', this.syncState);
+  }
+
+  componentWillUnmount() {
+    $.unsubscribe('.count');
+  }
+
+  render() {
+    return (
+      <div className='pswp__button pswp__button--vote-progress'>
+        <CircularProgress
+          current={this.state.voteCount}
+          max={this.state.maxVotes}
+          theme='gallery-contest'
+          tooltip={osu.trans('contest.voting.progress._', {
+            max: this.state.maxVotes,
+            used: this.state.voteCount,
+          })}
+        />
+      </div>
+    );
+  }
+
+  private getVoteState = () => {
+    const voteSummary = this.voteSummary();
+    const data = voteSummary.dataset.contestVoteSummary;
+
+    if (data === undefined) {
+      return { maxVotes: 0, voteCount: 0 };
+    }
+
+    const voteState: State = JSON.parse(data);
+    return voteState;
+  };
```

Comment:
The following is probably fine for now if we assume the element always has the expected dataset with the expected format.

The only difference being this also throws error if dataset is undefined instead of returning a blank "default" which is already kind of unexpected. And if we want to really account for such cases, the current one is missing check whether the dataset is in expected format...

```suggestion
  private getVoteState = (): State => JSON.parse(this.voteSummary().dataset.contestVoteSummary ?? '');
```

Reply:
Okay, I get it.